// n8n-core/dist/node-execution-context/base-execute-context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseExecuteContext = void 0;
const lodash_1 = require("lodash");
const n8n_workflow_1 = require("n8n-workflow");
const typedi_1 = require("typedi");
const BinaryData_service_1 = require("../BinaryData/BinaryData.service");
const node_execution_context_1 = require("./node-execution-context");
class BaseExecuteContext extends node_execution_context_1.NodeExecutionContext {
    constructor(workflow, node, additionalData, mode, runExecutionData, runIndex, connectionInputData, inputData, executeData, abortSignal) {
        super(workflow, node, additionalData, mode, runExecutionData, runIndex);
        this.runExecutionData = runExecutionData;
        this.connectionInputData = connectionInputData;
        this.inputData = inputData;
        this.executeData = executeData;
        this.abortSignal = abortSignal;
        this.binaryDataService = typedi_1.Container.get(BinaryData_service_1.BinaryDataService);
    }
    getExecutionCancelSignal() {
        return this.abortSignal;
    }
    onExecutionCancellation(handler) {
        const fn = () => {
            this.abortSignal?.removeEventListener('abort', fn);
            handler();
        };
        this.abortSignal?.addEventListener('abort', fn);
    }
    getExecuteData() {
        return this.executeData;
    }
    setMetadata(metadata) {
        this.executeData.metadata = {
            ...(this.executeData.metadata ?? {}),
            ...metadata,
        };
    }
    getContext(type) {
        return n8n_workflow_1.NodeHelpers.getContext(this.runExecutionData, type, this.node);
    }
    continueOnFail() {
        const onError = (0, lodash_1.get)(this.node, 'onError', undefined);
        if (onError === undefined) {
            return (0, lodash_1.get)(this.node, 'continueOnFail', false);
        }
        return ['continueRegularOutput', 'continueErrorOutput'].includes(onError);
    }
    async getCredentials(type, itemIndex) {
        return await this._getCredentials(type, this.executeData, this.connectionInputData, itemIndex);
    }
    async putExecutionToWait(waitTill) {
        this.runExecutionData.waitTill = waitTill;
        if (this.additionalData.setExecutionStatus) {
            this.additionalData.setExecutionStatus('waiting');
        }
    }
    async executeWorkflow(workflowInfo, inputData, parentCallbackManager, options) {
        const result = await this.additionalData.executeWorkflow(workflowInfo, this.additionalData, {
            ...options,
            parentWorkflowId: this.workflow.id,
            inputData,
            parentWorkflowSettings: this.workflow.settings,
            node: this.node,
            parentCallbackManager,
        });
        if (result.waitTill) {
            await this.putExecutionToWait(n8n_workflow_1.WAIT_INDEFINITELY);
        }
        const data = await this.binaryDataService.duplicateBinaryData(this.workflow.id, this.additionalData.executionId, result.data);
        return { ...result, data };
    }
    getInputItems(inputIndex, connectionType) {
        const inputData = this.inputData[connectionType];
        if (inputData.length < inputIndex) {
            throw new n8n_workflow_1.ApplicationError('Could not get input with given index', {
                extra: { inputIndex, connectionType },
            });
        }
        const allItems = inputData[inputIndex];
        if (allItems === null) {
            throw new n8n_workflow_1.ApplicationError('Input index was not set', {
                extra: { inputIndex, connectionType },
            });
        }
        return allItems;
    }
    getNodeInputs() {
        const nodeType = this.workflow.nodeTypes.getByNameAndVersion(this.node.type, this.node.typeVersion);
        return n8n_workflow_1.NodeHelpers.getNodeInputs(this.workflow, this.node, nodeType.description).map((input) => typeof input === 'string' ? { type: input } : input);
    }
    getNodeOutputs() {
        const nodeType = this.workflow.nodeTypes.getByNameAndVersion(this.node.type, this.node.typeVersion);
        return n8n_workflow_1.NodeHelpers.getNodeOutputs(this.workflow, this.node, nodeType.description).map((output) => (typeof output === 'string' ? { type: output } : output));
    }
    getInputSourceData(inputIndex = 0, connectionType = "main") {
        if (this.executeData?.source === null) {
            throw new n8n_workflow_1.ApplicationError('Source data is missing');
        }
        return this.executeData.source[connectionType][inputIndex];
    }
    getWorkflowDataProxy(itemIndex) {
        return new n8n_workflow_1.WorkflowDataProxy(this.workflow, this.runExecutionData, this.runIndex, itemIndex, this.node.name, this.connectionInputData, {}, this.mode, this.additionalKeys, this.executeData).getDataProxy();
    }
    sendMessageToUI(...args) {
        if (this.mode !== 'manual') {
            return;
        }
        try {
            if (this.additionalData.sendDataToUI) {
                args = args.map((arg) => {
                    if (arg.isLuxonDateTime && arg.invalidReason)
                        return { ...arg };
                    if (arg.isLuxonDateTime)
                        return new Date(arg.ts).toString();
                    if (arg instanceof Date)
                        return arg.toString();
                    return arg;
                });
                this.additionalData.sendDataToUI('sendConsoleMessage', {
                    source: `[Node: "${this.node.name}"]`,
                    messages: args,
                });
            }
        }
        catch (error) {
            this.logger.warn(`There was a problem sending message to UI: ${error.message}`);
        }
    }
    logAiEvent(eventName, msg) {
        return this.additionalData.logAiEvent(eventName, {
            executionId: this.additionalData.executionId ?? 'unsaved-execution',
            nodeName: this.node.name,
            workflowName: this.workflow.name ?? 'Unnamed workflow',
            nodeType: this.node.type,
            workflowId: this.workflow.id ?? 'unsaved-workflow',
            msg,
        });
    }
}
exports.BaseExecuteContext = BaseExecuteContext;
