// n8n-core/dist/node-execution-context/execute-single-context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExecuteSingleContext = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const NodeExecuteFunctions_1 = require("../NodeExecuteFunctions");
const base_execute_context_1 = require("./base-execute-context");
class ExecuteSingleContext extends base_execute_context_1.BaseExecuteContext {
    constructor(workflow, node, additionalData, mode, runExecutionData, runIndex, connectionInputData, inputData, itemIndex, executeData, abortSignal) {
        super(workflow, node, additionalData, mode, runExecutionData, runIndex, connectionInputData, inputData, executeData, abortSignal);
        this.itemIndex = itemIndex;
        this.helpers = {
            createDeferredPromise: n8n_workflow_1.createDeferredPromise,
            returnJsonArray: NodeExecuteFunctions_1.returnJsonArray,
            ...(0, NodeExecuteFunctions_1.getRequestHelperFunctions)(workflow, node, additionalData, runExecutionData, connectionInputData),
            ...(0, NodeExecuteFunctions_1.getBinaryHelperFunctions)(additionalData, workflow.id),
            assertBinaryData: (propertyName, inputIndex = 0) => (0, NodeExecuteFunctions_1.assertBinaryData)(inputData, node, itemIndex, propertyName, inputIndex),
            getBinaryDataBuffer: async (propertyName, inputIndex = 0) => await (0, NodeExecuteFunctions_1.getBinaryDataBuffer)(inputData, itemIndex, propertyName, inputIndex),
        };
    }
    evaluateExpression(expression, itemIndex = this.itemIndex) {
        return super.evaluateExpression(expression, itemIndex);
    }
    getInputData(inputIndex = 0, connectionType = "main") {
        if (!this.inputData.hasOwnProperty(connectionType)) {
            return { json: {} };
        }
        const allItems = super.getInputItems(inputIndex, connectionType);
        const data = allItems?.[this.itemIndex];
        if (data === undefined) {
            throw new n8n_workflow_1.ApplicationError('Value of input with given index was not set', {
                extra: { inputIndex, connectionType, itemIndex: this.itemIndex },
            });
        }
        return data;
    }
    getItemIndex() {
        return this.itemIndex;
    }
    getNodeParameter(parameterName, fallbackValue, options) {
        return this._getNodeParameter(parameterName, this.itemIndex, fallbackValue, options);
    }
    async getCredentials(type) {
        return await super.getCredentials(type, this.itemIndex);
    }
    getWorkflowDataProxy() {
        return super.getWorkflowDataProxy(this.itemIndex);
    }
}
exports.ExecuteSingleContext = ExecuteSingleContext;
