// n8n-core/dist/node-execution-context/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAdditionalKeys = exports.WebhookContext = exports.TriggerContext = exports.SupplyDataContext = exports.PollContext = exports.LoadOptionsContext = exports.HookContext = exports.ExecuteSingleContext = exports.ExecuteContext = void 0;
var execute_context_1 = require("./execute-context");
Object.defineProperty(exports, "ExecuteContext", { enumerable: true, get: function () { return execute_context_1.ExecuteContext; } });
var execute_single_context_1 = require("./execute-single-context");
Object.defineProperty(exports, "ExecuteSingleContext", { enumerable: true, get: function () { return execute_single_context_1.ExecuteSingleContext; } });
var hook_context_1 = require("./hook-context");
Object.defineProperty(exports, "HookContext", { enumerable: true, get: function () { return hook_context_1.HookContext; } });
var load_options_context_1 = require("./load-options-context");
Object.defineProperty(exports, "LoadOptionsContext", { enumerable: true, get: function () { return load_options_context_1.LoadOptionsContext; } });
var poll_context_1 = require("./poll-context");
Object.defineProperty(exports, "PollContext", { enumerable: true, get: function () { return poll_context_1.PollContext; } });
var supply_data_context_1 = require("./supply-data-context");
Object.defineProperty(exports, "SupplyDataContext", { enumerable: true, get: function () { return supply_data_context_1.SupplyDataContext; } });
var trigger_context_1 = require("./trigger-context");
Object.defineProperty(exports, "TriggerContext", { enumerable: true, get: function () { return trigger_context_1.TriggerContext; } });
var webhook_context_1 = require("./webhook-context");
Object.defineProperty(exports, "WebhookContext", { enumerable: true, get: function () { return webhook_context_1.WebhookContext; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "getAdditionalKeys", { enumerable: true, get: function () { return utils_1.getAdditionalKeys; } });
