// n8n-core/dist/node-execution-context/trigger-context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TriggerContext = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const NodeExecuteFunctions_1 = require("../NodeExecuteFunctions");
const node_execution_context_1 = require("./node-execution-context");
const throwOnEmit = () => {
    throw new n8n_workflow_1.ApplicationError('Overwrite TriggerContext.emit function');
};
const throwOnEmitError = () => {
    throw new n8n_workflow_1.ApplicationError('Overwrite TriggerContext.emitError function');
};
class TriggerContext extends node_execution_context_1.NodeExecutionContext {
    constructor(workflow, node, additionalData, mode, activation, emit = throwOnEmit, emitError = throwOnEmitError) {
        super(workflow, node, additionalData, mode);
        this.activation = activation;
        this.emit = emit;
        this.emitError = emitError;
        this.helpers = {
            createDeferredPromise: n8n_workflow_1.createDeferredPromise,
            returnJsonArray: NodeExecuteFunctions_1.returnJsonArray,
            ...(0, NodeExecuteFunctions_1.getSSHTunnelFunctions)(),
            ...(0, NodeExecuteFunctions_1.getRequestHelperFunctions)(workflow, node, additionalData),
            ...(0, NodeExecuteFunctions_1.getBinaryHelperFunctions)(additionalData, workflow.id),
            ...(0, NodeExecuteFunctions_1.getSchedulingFunctions)(workflow),
        };
    }
    getActivationMode() {
        return this.activation;
    }
    async getCredentials(type) {
        return await this._getCredentials(type);
    }
}
exports.TriggerContext = TriggerContext;
