// n8n-core/dist/DirectoryLoader.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LazyPackageDirectoryLoader = exports.PackageDirectoryLoader = exports.CustomDirectoryLoader = exports.DirectoryLoader = void 0;
const fast_glob_1 = __importDefault(require("fast-glob"));
const uniqBy_1 = __importDefault(require("lodash/uniqBy"));
const n8n_workflow_1 = require("n8n-workflow");
const node_fs_1 = require("node:fs");
const promises_1 = require("node:fs/promises");
const path = __importStar(require("path"));
const ClassLoader_1 = require("./ClassLoader");
const Constants_1 = require("./Constants");
const unrecognized_credential_type_error_1 = require("./errors/unrecognized-credential-type.error");
const unrecognized_node_type_error_1 = require("./errors/unrecognized-node-type.error");
function toJSON() {
    return {
        ...this,
        authenticate: typeof this.authenticate === 'function' ? {} : this.authenticate,
    };
}
class DirectoryLoader {
    constructor(directory, excludeNodes = [], includeNodes = []) {
        this.directory = directory;
        this.excludeNodes = excludeNodes;
        this.includeNodes = includeNodes;
        this.isLazyLoaded = false;
        this.loadedNodes = [];
        this.nodeTypes = {};
        this.credentialTypes = {};
        this.known = { nodes: {}, credentials: {} };
        this.types = { nodes: [], credentials: [] };
        this.nodesByCredential = {};
    }
    reset() {
        this.loadedNodes = [];
        this.nodeTypes = {};
        this.credentialTypes = {};
        this.known = { nodes: {}, credentials: {} };
        this.types = { nodes: [], credentials: [] };
    }
    resolvePath(file) {
        return path.resolve(this.directory, file);
    }
    loadClass(sourcePath) {
        const filePath = this.resolvePath(sourcePath);
        const [className] = path.parse(sourcePath).name.split('.');
        try {
            return (0, ClassLoader_1.loadClassInIsolation)(filePath, className);
        }
        catch (error) {
            throw error instanceof TypeError
                ? new n8n_workflow_1.ApplicationError('Class could not be found. Please check if the class is named correctly.', { extra: { className } })
                : error;
        }
    }
    loadNodeFromFile(filePath) {
        const tempNode = this.loadClass(filePath);
        this.addCodex(tempNode, filePath);
        const nodeType = tempNode.description.name;
        const fullNodeType = `${this.packageName}.${nodeType}`;
        if (this.includeNodes.length && !this.includeNodes.includes(fullNodeType)) {
            return;
        }
        if (this.excludeNodes.includes(fullNodeType)) {
            return;
        }
        this.fixIconPaths(tempNode.description, filePath);
        let nodeVersion = 1;
        if ('nodeVersions' in tempNode) {
            for (const versionNode of Object.values(tempNode.nodeVersions)) {
                this.fixIconPaths(versionNode.description, filePath);
            }
            for (const version of Object.values(tempNode.nodeVersions)) {
                this.addLoadOptionsMethods(version);
                n8n_workflow_1.NodeHelpers.applySpecialNodeParameters(version);
            }
            const currentVersionNode = tempNode.nodeVersions[tempNode.currentVersion];
            this.addCodex(currentVersionNode, filePath);
            nodeVersion = tempNode.currentVersion;
            if (currentVersionNode.hasOwnProperty('executeSingle')) {
                throw new n8n_workflow_1.ApplicationError('"executeSingle" has been removed. Please update the code of this node to use "execute" instead.', { extra: { nodeType: fullNodeType } });
            }
        }
        else {
            this.addLoadOptionsMethods(tempNode);
            n8n_workflow_1.NodeHelpers.applySpecialNodeParameters(tempNode);
            nodeVersion = Array.isArray(tempNode.description.version)
                ? tempNode.description.version.slice(-1)[0]
                : tempNode.description.version;
        }
        this.known.nodes[nodeType] = {
            className: tempNode.constructor.name,
            sourcePath: filePath,
        };
        this.nodeTypes[nodeType] = {
            type: tempNode,
            sourcePath: filePath,
        };
        this.loadedNodes.push({
            name: nodeType,
            version: nodeVersion,
        });
        this.getVersionedNodeTypeAll(tempNode).forEach(({ description }) => {
            this.types.nodes.push(description);
        });
        for (const credential of this.getCredentialsForNode(tempNode)) {
            if (!this.nodesByCredential[credential.name]) {
                this.nodesByCredential[credential.name] = [];
            }
            this.nodesByCredential[credential.name].push(nodeType);
        }
    }
    getNode(nodeType) {
        const { nodeTypes, known: { nodes: knownNodes }, } = this;
        if (!(nodeType in nodeTypes) && nodeType in knownNodes) {
            const { sourcePath } = knownNodes[nodeType];
            this.loadNodeFromFile(sourcePath);
        }
        if (nodeType in nodeTypes) {
            return nodeTypes[nodeType];
        }
        throw new unrecognized_node_type_error_1.UnrecognizedNodeTypeError(this.packageName, nodeType);
    }
    loadCredentialFromFile(filePath) {
        const tempCredential = this.loadClass(filePath);
        Object.assign(tempCredential, { toJSON });
        this.fixIconPaths(tempCredential, filePath);
        const credentialType = tempCredential.name;
        this.known.credentials[credentialType] = {
            className: tempCredential.constructor.name,
            sourcePath: filePath,
            extends: tempCredential.extends,
            supportedNodes: this.nodesByCredential[credentialType],
        };
        this.credentialTypes[credentialType] = {
            type: tempCredential,
            sourcePath: filePath,
        };
        this.types.credentials.push(tempCredential);
    }
    getCredential(credentialType) {
        const { credentialTypes, known: { credentials: knownCredentials }, } = this;
        if (!(credentialType in credentialTypes) && credentialType in knownCredentials) {
            const { sourcePath } = knownCredentials[credentialType];
            this.loadCredentialFromFile(sourcePath);
        }
        if (credentialType in credentialTypes) {
            return credentialTypes[credentialType];
        }
        throw new unrecognized_credential_type_error_1.UnrecognizedCredentialTypeError(credentialType);
    }
    getCredentialsForNode(object) {
        if ('nodeVersions' in object) {
            const credentials = Object.values(object.nodeVersions).flatMap(({ description }) => description.credentials ?? []);
            return (0, uniqBy_1.default)(credentials, 'name');
        }
        return object.description.credentials ?? [];
    }
    getVersionedNodeTypeAll(object) {
        if ('nodeVersions' in object) {
            const nodeVersions = Object.values(object.nodeVersions).map((element) => {
                element.description.name = object.description.name;
                element.description.codex = object.description.codex;
                return element;
            });
            return (0, uniqBy_1.default)(nodeVersions.reverse(), (node) => {
                const { version } = node.description;
                return Array.isArray(version) ? version.join(',') : version.toString();
            });
        }
        return [object];
    }
    getCodex(filePath) {
        const codexFilePath = this.resolvePath(`${filePath}on`);
        const { categories, subcategories, resources: { primaryDocumentation, credentialDocumentation }, alias, } = module.require(codexFilePath);
        return {
            ...(categories && { categories }),
            ...(subcategories && { subcategories }),
            ...(alias && { alias }),
            resources: {
                primaryDocumentation,
                credentialDocumentation,
            },
        };
    }
    addCodex(node, filePath) {
        const isCustom = this.packageName === 'CUSTOM';
        try {
            let codex;
            if (!isCustom) {
                codex = node.description.codex;
            }
            if (codex === undefined) {
                codex = this.getCodex(filePath);
            }
            if (isCustom) {
                codex.categories = codex.categories
                    ? codex.categories.concat(Constants_1.CUSTOM_NODES_CATEGORY)
                    : [Constants_1.CUSTOM_NODES_CATEGORY];
            }
            node.description.codex = codex;
        }
        catch {
            n8n_workflow_1.LoggerProxy.debug(`No codex available for: ${node.description.name}`);
            if (isCustom) {
                node.description.codex = {
                    categories: [Constants_1.CUSTOM_NODES_CATEGORY],
                };
            }
        }
    }
    addLoadOptionsMethods(node) {
        if (node?.methods?.loadOptions) {
            node.description.__loadOptionsMethods = Object.keys(node.methods.loadOptions);
        }
    }
    getIconPath(icon, filePath) {
        const iconPath = path.join(path.dirname(filePath), icon.replace('file:', ''));
        return `icons/${this.packageName}/${iconPath}`;
    }
    fixIconPaths(obj, filePath) {
        const { icon } = obj;
        if (!icon)
            return;
        if (typeof icon === 'string') {
            if (icon.startsWith('file:')) {
                obj.iconUrl = this.getIconPath(icon, filePath);
                obj.icon = undefined;
            }
        }
        else if (icon.light.startsWith('file:') && icon.dark.startsWith('file:')) {
            obj.iconUrl = {
                light: this.getIconPath(icon.light, filePath),
                dark: this.getIconPath(icon.dark, filePath),
            };
            obj.icon = undefined;
        }
    }
}
exports.DirectoryLoader = DirectoryLoader;
class CustomDirectoryLoader extends DirectoryLoader {
    constructor() {
        super(...arguments);
        this.packageName = 'CUSTOM';
    }
    async loadAll() {
        const nodes = await (0, fast_glob_1.default)('**/*.node.js', {
            cwd: this.directory,
            absolute: true,
        });
        for (const nodePath of nodes) {
            this.loadNodeFromFile(nodePath);
        }
        const credentials = await (0, fast_glob_1.default)('**/*.credentials.js', {
            cwd: this.directory,
            absolute: true,
        });
        for (const credentialPath of credentials) {
            this.loadCredentialFromFile(credentialPath);
        }
    }
}
exports.CustomDirectoryLoader = CustomDirectoryLoader;
class PackageDirectoryLoader extends DirectoryLoader {
    constructor() {
        super(...arguments);
        this.packageJson = this.readJSONSync('package.json');
        this.packageName = this.packageJson.name;
    }
    async loadAll() {
        const { n8n } = this.packageJson;
        if (!n8n)
            return;
        const { nodes, credentials } = n8n;
        if (Array.isArray(nodes)) {
            for (const nodePath of nodes) {
                this.loadNodeFromFile(nodePath);
            }
        }
        if (Array.isArray(credentials)) {
            for (const credentialPath of credentials) {
                this.loadCredentialFromFile(credentialPath);
            }
        }
        this.inferSupportedNodes();
        n8n_workflow_1.LoggerProxy.debug(`Loaded all credentials and nodes from ${this.packageName}`, {
            credentials: credentials?.length ?? 0,
            nodes: nodes?.length ?? 0,
        });
    }
    inferSupportedNodes() {
        const knownCredentials = this.known.credentials;
        for (const { type: credentialType } of Object.values(this.credentialTypes)) {
            const supportedNodes = knownCredentials[credentialType.name].supportedNodes ?? [];
            if (supportedNodes.length > 0 && credentialType.httpRequestNode) {
                credentialType.httpRequestNode.hidden = true;
            }
            credentialType.supportedNodes = supportedNodes;
            if (!credentialType.iconUrl && !credentialType.icon) {
                for (const supportedNode of supportedNodes) {
                    const nodeDescription = this.nodeTypes[supportedNode]?.type.description;
                    if (!nodeDescription)
                        continue;
                    if (nodeDescription.icon) {
                        credentialType.icon = nodeDescription.icon;
                        credentialType.iconColor = nodeDescription.iconColor;
                        break;
                    }
                    if (nodeDescription.iconUrl) {
                        credentialType.iconUrl = nodeDescription.iconUrl;
                        break;
                    }
                }
            }
        }
    }
    parseJSON(fileString, filePath) {
        try {
            return (0, n8n_workflow_1.jsonParse)(fileString);
        }
        catch (error) {
            throw new n8n_workflow_1.ApplicationError('Failed to parse JSON', { extra: { filePath } });
        }
    }
    readJSONSync(file) {
        const filePath = this.resolvePath(file);
        const fileString = (0, node_fs_1.readFileSync)(filePath, 'utf8');
        return this.parseJSON(fileString, filePath);
    }
    async readJSON(file) {
        const filePath = this.resolvePath(file);
        const fileString = await (0, promises_1.readFile)(filePath, 'utf8');
        return this.parseJSON(fileString, filePath);
    }
}
exports.PackageDirectoryLoader = PackageDirectoryLoader;
class LazyPackageDirectoryLoader extends PackageDirectoryLoader {
    async loadAll() {
        try {
            this.known.nodes = await this.readJSON('dist/known/nodes.json');
            this.known.credentials = await this.readJSON('dist/known/credentials.json');
            this.types.nodes = await this.readJSON('dist/types/nodes.json');
            this.types.credentials = await this.readJSON('dist/types/credentials.json');
            if (this.includeNodes.length) {
                const allowedNodes = {};
                for (const fullNodeType of this.includeNodes) {
                    const [packageName, nodeType] = fullNodeType.split('.');
                    if (packageName === this.packageName && nodeType in this.known.nodes) {
                        allowedNodes[nodeType] = this.known.nodes[nodeType];
                    }
                }
                this.known.nodes = allowedNodes;
                this.types.nodes = this.types.nodes.filter((nodeType) => this.includeNodes.includes(nodeType.name));
            }
            if (this.excludeNodes.length) {
                for (const fullNodeType of this.excludeNodes) {
                    const [packageName, nodeType] = fullNodeType.split('.');
                    if (packageName === this.packageName) {
                        delete this.known.nodes[nodeType];
                    }
                }
                this.types.nodes = this.types.nodes.filter((nodeType) => !this.excludeNodes.includes(nodeType.name));
            }
            n8n_workflow_1.LoggerProxy.debug(`Lazy-loading nodes and credentials from ${this.packageJson.name}`, {
                nodes: this.types.nodes?.length ?? 0,
                credentials: this.types.credentials?.length ?? 0,
            });
            this.isLazyLoaded = true;
            return;
        }
        catch {
            n8n_workflow_1.LoggerProxy.debug("Can't enable lazy-loading");
            await super.loadAll();
        }
    }
}
exports.LazyPackageDirectoryLoader = LazyPackageDirectoryLoader;
