// n8n-core/dist/InstanceSettings.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InstanceSettings = void 0;
const crypto_1 = require("crypto");
const fs_1 = require("fs");
const n8n_workflow_1 = require("n8n-workflow");
const nanoid_1 = require("nanoid");
const path_1 = __importDefault(require("path"));
const typedi_1 = require("typedi");
const InstanceSettingsConfig_1 = require("./InstanceSettingsConfig");
const nanoid = (0, nanoid_1.customAlphabet)(n8n_workflow_1.ALPHABET, 16);
const inTest = process.env.NODE_ENV === 'test';
let InstanceSettings = class InstanceSettings {
    constructor(config) {
        this.config = config;
        this.userHome = this.getUserHome();
        this.n8nFolder = path_1.default.join(this.userHome, '.n8n');
        this.staticCacheDir = path_1.default.join(this.userHome, '.cache/n8n/public');
        this.customExtensionDir = path_1.default.join(this.n8nFolder, 'custom');
        this.nodesDownloadDir = path_1.default.join(this.n8nFolder, 'nodes');
        this.settingsFile = path_1.default.join(this.n8nFolder, 'config');
        this.enforceSettingsFilePermissions = this.loadEnforceSettingsFilePermissionsFlag();
        this.settings = this.loadOrCreate();
        this.instanceId = this.generateInstanceId();
        this.instanceRole = 'unset';
        const command = process.argv[2];
        this.instanceType = ['webhook', 'worker'].includes(command)
            ? command
            : 'main';
        this.hostId = `${this.instanceType}-${nanoid()}`;
    }
    get isLeader() {
        return this.instanceRole === 'leader';
    }
    markAsLeader() {
        this.instanceRole = 'leader';
    }
    get isFollower() {
        return this.instanceRole === 'follower';
    }
    markAsFollower() {
        this.instanceRole = 'follower';
    }
    get encryptionKey() {
        return this.settings.encryptionKey;
    }
    get tunnelSubdomain() {
        return this.settings.tunnelSubdomain;
    }
    update(newSettings) {
        this.save({ ...this.settings, ...newSettings });
    }
    getUserHome() {
        const homeVarName = process.platform === 'win32' ? 'USERPROFILE' : 'HOME';
        return process.env.N8N_USER_FOLDER ?? process.env[homeVarName] ?? process.cwd();
    }
    loadOrCreate() {
        if ((0, fs_1.existsSync)(this.settingsFile)) {
            const content = (0, fs_1.readFileSync)(this.settingsFile, 'utf8');
            this.ensureSettingsFilePermissions();
            const settings = (0, n8n_workflow_1.jsonParse)(content, {
                errorMessage: `Error parsing n8n-config file "${this.settingsFile}". It does not seem to be valid JSON.`,
            });
            if (!inTest)
                console.info(`User settings loaded from: ${this.settingsFile}`);
            const { encryptionKey, tunnelSubdomain } = settings;
            if (process.env.N8N_ENCRYPTION_KEY && encryptionKey !== process.env.N8N_ENCRYPTION_KEY) {
                throw new n8n_workflow_1.ApplicationError(`Mismatching encryption keys. The encryption key in the settings file ${this.settingsFile} does not match the N8N_ENCRYPTION_KEY env var. Please make sure both keys match. More information: https://docs.n8n.io/hosting/environment-variables/configuration-methods/#encryption-key`);
            }
            return { encryptionKey, tunnelSubdomain };
        }
        (0, fs_1.mkdirSync)(this.n8nFolder, { recursive: true });
        const encryptionKey = process.env.N8N_ENCRYPTION_KEY ?? (0, crypto_1.randomBytes)(24).toString('base64');
        const settings = { encryptionKey };
        this.save(settings);
        if (!inTest && !process.env.N8N_ENCRYPTION_KEY) {
            console.info(`No encryption key found - Auto-generated and saved to: ${this.settingsFile}`);
        }
        this.ensureSettingsFilePermissions();
        return settings;
    }
    generateInstanceId() {
        const { encryptionKey } = this;
        return (0, crypto_1.createHash)('sha256')
            .update(encryptionKey.slice(Math.round(encryptionKey.length / 2)))
            .digest('hex');
    }
    save(settings) {
        this.settings = settings;
        (0, fs_1.writeFileSync)(this.settingsFile, JSON.stringify(this.settings, null, '\t'), {
            mode: this.enforceSettingsFilePermissions.enforce ? 0o600 : undefined,
            encoding: 'utf-8',
        });
    }
    loadEnforceSettingsFilePermissionsFlag() {
        const { enforceSettingsFilePermissions } = this.config;
        const isEnvVarSet = !!process.env.N8N_ENFORCE_SETTINGS_FILE_PERMISSIONS;
        if (this.isWindows()) {
            if (isEnvVarSet) {
                console.warn('Ignoring N8N_ENFORCE_SETTINGS_FILE_PERMISSIONS as it is not supported on Windows.');
            }
            return {
                isSet: isEnvVarSet,
                enforce: false,
            };
        }
        return {
            isSet: isEnvVarSet,
            enforce: enforceSettingsFilePermissions,
        };
    }
    ensureSettingsFilePermissions() {
        if (this.enforceSettingsFilePermissions.isSet && !this.enforceSettingsFilePermissions.enforce) {
            return;
        }
        if (this.isWindows()) {
            return;
        }
        const permissionsResult = (0, n8n_workflow_1.toResult)(() => {
            const stats = (0, fs_1.statSync)(this.settingsFile);
            return stats.mode & 0o777;
        });
        if (!permissionsResult.ok) {
            console.warn(`Could not ensure settings file permissions: ${permissionsResult.error.message}. To skip this check, set N8N_ENFORCE_SETTINGS_FILE_PERMISSIONS=false.`);
            return;
        }
        const arePermissionsCorrect = permissionsResult.result === 0o600;
        if (arePermissionsCorrect) {
            return;
        }
        if (!this.enforceSettingsFilePermissions.isSet) {
            console.warn(`Permissions 0${permissionsResult.result.toString(8)} for n8n settings file ${this.settingsFile} are too wide. This is ignored for now, but in the future n8n will attempt to change the permissions automatically. To automatically enforce correct permissions now set N8N_ENFORCE_SETTINGS_FILE_PERMISSIONS=true (recommended), or turn this check off set N8N_ENFORCE_SETTINGS_FILE_PERMISSIONS=false.`);
            return;
        }
        if (this.enforceSettingsFilePermissions.enforce) {
            console.warn(`Permissions 0${permissionsResult.result.toString(8)} for n8n settings file ${this.settingsFile} are too wide. Changing permissions to 0600..`);
            const chmodResult = (0, n8n_workflow_1.toResult)(() => (0, fs_1.chmodSync)(this.settingsFile, 0o600));
            if (!chmodResult.ok) {
                console.warn(`Could not enforce settings file permissions: ${chmodResult.error.message}. To skip this check, set N8N_ENFORCE_SETTINGS_FILE_PERMISSIONS=false.`);
            }
        }
    }
    isWindows() {
        return process.platform === 'win32';
    }
};
exports.InstanceSettings = InstanceSettings;
exports.InstanceSettings = InstanceSettings = __decorate([
    (0, typedi_1.Service)(),
    __metadata("design:paramtypes", [InstanceSettingsConfig_1.InstanceSettingsConfig])
], InstanceSettings);
